crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -87,23 +87,22 @@ fn all_negative_narrowing_constraints_for_expression<'db>(
 ///
 /// The `classinfo` argument can be a class literal, a tuple of (tuples of) class literals. PEP 604
 /// union types are not yet supported. Returns `None` if the `classinfo` argument has a wrong type.
-fn generate_classinfo_constraint<'db, F>(
+fn generate_classinfo_constraint<'db>(
     db: &'db dyn Db,
     classinfo: &Type<'db>,
-    to_constraint: F,
-) -> Option<Type<'db>>
-where
-    F: Fn(ClassLiteralType<'db>) -> Type<'db> + Copy,
-{
+    constraint_fn: KnownConstraintFunction,
+) -> Option<Type<'db>> {
     match classinfo {
         Type::Tuple(tuple) => {
             let mut builder = UnionBuilder::new(db);
             for element in tuple.elements(db) {
-                builder = builder.add(generate_classinfo_constraint(db, element, to_constraint)?);
+                builder = builder.add(generate_classinfo_constraint(db, element, constraint_fn)?);
             }
             Some(builder.build())
         }
-        Type::ClassLiteral(class_literal_type) => Some(to_constraint(*class_literal_type)),
+        Type::ClassLiteral(ClassLiteralType { class }) => {
```

Comment:
Orthogonal to this PR, but I think we should probably also match `Type::SubclassOf` here? I would expect code like this to narrow:

```py
def f(instance: A | None, ty: type[A]):
    if isinstance(instance, ty):
        reveal_type(instance)  # revealed: A
```

Reply:
Nice catch -- yes, I think you're right! We probably added this logic before we had the `Type::SubclassOf` variant. I won't fix it in this PR, though, as it _is_ orthogonal to what this PR is doing, and it'll probably complicate rebasing #15272 onto this branch

Reply:
Oh, actually, addressing your other review comment already throws away any chance I have of a clean rebase of that PR branch onto this one, so I may as well fix it here 😆